doc/sphinx/proofs/automatic-tactics/auto.rst

```diff
@@ -133,81 +116,84 @@ Programmable proof search
          Goal forall P:nat -> Prop, P 0 -> exists n, P n.
          eauto.
 
-      Note that ``ex_intro`` should be declared as a hint.
+      `ex_intro` is declared as a hint so the proof succeeds.
 
+   .. seealso:: :ref:`thehintsdatabasesforautoandeauto`
 
-   .. tacv:: {? info_}eauto {? @natural} {? using {+ @qualid}} {? with {+ @ident}}
+   .. tacn:: info_eauto {? @int_or_var } {? @auto_using } {? @hintbases }
 
-      The various options for :tacn:`eauto` are the same as for :tacn:`auto`.
+      The various options for :tacn:`info_eauto` are the same as for :tacn:`info_auto`.
 
    :tacn:`eauto` also obeys the following flags:
 
    .. flag:: Info Eauto
              Debug Eauto
       :undocumented:
 
-   .. seealso:: :ref:`The Hints Databases for auto and eauto <thehintsdatabasesforautoandeauto>`
-
+   .. tacn:: debug eauto {? @int_or_var } {? @auto_using } {? @hintbases }
 
-.. tacn:: autounfold with {+ @ident}
-   :name: autounfold
+      Behaves like :tacn:`eauto` but shows the tactics it tries to solve the goal,
+      including failing paths.
 
-   This tactic unfolds constants that were declared through a :cmd:`Hint Unfold`
-   in the given databases.
+   .. tacn::bfs eauto
 
-.. tacv:: autounfold with {+ @ident} in @goal_occurrences
+      Like :tacn:`eauto`, but does a breadth-first search (i.e. the tactic
+      checks all the relevant hints to see if one solves the goal before
+      recurring).
```

Comment:
breadth-first search is more than this (it explores all "paths of length 2" before going to "paths of length 3"). Anyway, what about linking to [Wikipedia](https://en.wikipedia.org/wiki/Breadth-first_search) instead of attempting to explain this?

Reply:
Got it.  I had some doubt that this was doing a full bfs.  AFAIK, the kernel, monad infrastructure, Ltac, etc. are all designed around dfs and backtracking.  For efficiency, you'd want to save the resulting proof state for each successful hint at each level rather than repeating application of the hint (easy to code).  I'll have to look at that code sometime.

One thing I'd like to experiment with at some point is guiding the search in part by metrics on the proof such as the length of the goal conclusion(s), keeping only the best N results at any point.  That may solve some goals more efficiently and/or make useful progress for the user.

Reply:
BTW, there are several implementations of proof search and the plan was to get rid of the legacy `auto` and `eauto` backend in favor of the most recently written `typeclasses eauto` which also has a bfs strategy BTW (while keeping the tactics of course). See #721.

Reply:
Ah, you worked on that!  We'll have to chat about it sometime.